File: src/Starfield2026.MiniToolbox/src/MiniToolbox.Core/Bntx/BinaryData/BinaryDataWriter.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Reflection;
using System.Text;
using MiniToolbox.Core.Bntx;

namespace MiniToolbox.Core.Bntx;

[DebuggerDisplay("BinaryDataWriter, Position={Position}")]
public class BinaryDataWriter : BinaryWriter
{
	private ByteOrder _byteOrder;

	public ByteOrder ByteOrder
	{
		get
		{
			return _byteOrder;
		}
		set
		{
			_byteOrder = value;
			NeedsReversion = _byteOrder != ByteOrderHelper.SystemByteOrder;
		}
	}

	public Encoding Encoding { get; private set; }

	public bool NeedsReversion { get; private set; }

	public long Position
	{
		get
		{
			return BaseStream.Position;
		}
		set
		{
			BaseStream.Position = value;
		}
	}

	public BinaryDataWriter(Stream output)
		: this(output, new UTF8Encoding(), leaveOpen: false)
	{
	}

	public BinaryDataWriter(Stream output, bool leaveOpen)
		: this(output, new UTF8Encoding(), leaveOpen)
	{
	}

	public BinaryDataWriter(Stream output, Encoding encoding)
		: this(output, encoding, leaveOpen: false)
	{
	}

	public BinaryDataWriter(Stream output, Encoding encoding, bool leaveOpen)
		: base(output, encoding, leaveOpen)
	{
		Encoding = encoding;
		ByteOrder = ByteOrderHelper.SystemByteOrder;
	}

	public void Align(int alignment)
	{
		Seek((-Position % alignment + alignment) % alignment);
	}

	public Offset ReserveOffset()
	{
		return new Offset(this);
	}

	public Offset[] ReserveOffset(int count)
	{
		Offset[] array = new Offset[count];
		for (int i = 0; i < count; i++)
		{
			array[i] = ReserveOffset();
		}
		return array;
	}

	public long Seek(long offset)
	{
		return Seek(offset, SeekOrigin.Current);
	}

	public long Seek(long offset, SeekOrigin origin)
	{
		return BaseStream.Seek(offset, origin);
	}

	public SeekTask TemporarySeek()
	{
		return TemporarySeek(0L, SeekOrigin.Current);
	}

	public SeekTask TemporarySeek(long offset)
	{
		return TemporarySeek(offset, SeekOrigin.Current);
	}

	public SeekTask TemporarySeek(long offset, SeekOrigin origin)
	{
		return new SeekTask(BaseStream, offset, origin);
	}

	public void Write(bool value, BinaryBooleanFormat format)
	{
		switch (format)
		{
		case BinaryBooleanFormat.NonZeroByte:
			base.Write(value);
			break;
		case BinaryBooleanFormat.NonZeroWord:
			Write((short)(value ? 1 : 0));
			break;
		case BinaryBooleanFormat.NonZeroDword:
			Write(value ? 1 : 0);
			break;
		default:
			throw new ArgumentOutOfRangeException("format", "The specified binary boolean format is invalid.");
		}
	}

	public void Write(IEnumerable<bool> values)
	{
		foreach (bool value in values)
		{
			Write(value);
		}
	}

	public void Write(IEnumerable<bool> values, BinaryBooleanFormat format)
	{
		foreach (bool value in values)
		{
			Write(value, format);
		}
	}

	public void Write(DateTime value)
	{
		Write(value, BinaryDateTimeFormat.NetTicks);
	}

	public void Write(DateTime value, BinaryDateTimeFormat format)
	{
		switch (format)
		{
		case BinaryDateTimeFormat.CTime:
			Write((uint)(new DateTime(1970, 1, 1) - value.ToLocalTime()).TotalSeconds);
			break;
		case BinaryDateTimeFormat.NetTicks:
			Write(value.Ticks);
			break;
		default:
			throw new ArgumentOutOfRangeException("format", "The specified binary date time format is invalid.");
		}
	}

	public void Write(IEnumerable<DateTime> values)
	{
		foreach (DateTime value in values)
		{
			Write(value, BinaryDateTimeFormat.NetTicks);
		}
	}

	public void Write(IEnumerable<DateTime> values, BinaryDateTimeFormat format)
	{
		foreach (DateTime value in values)
		{
			Write(value, format);
		}
	}

	public override void Write(decimal value)
	{
		if (NeedsReversion)
		{
			byte[] bytes = DecimalToBytes(value);
			WriteReversed(bytes);
		}
		else
		{
			base.Write(value);
		}
	}

	public void Write(IEnumerable<decimal> values)
	{
		foreach (decimal value in values)
		{
			Write(value);
		}
	}

	public override void Write(double value)
	{
		if (NeedsReversion)
		{
			byte[] bytes = BitConverter.GetBytes(value);
			WriteReversed(bytes);
		}
		else
		{
			base.Write(value);
		}
	}

	public void Write(IEnumerable<double> values)
	{
		foreach (double value in values)
		{
			Write(value);
		}
	}

	public void Write<T>(T value, bool strict) where T : struct, IComparable, IFormattable
	{
		WriteEnum(typeof(T), value, strict);
	}

	public void Write<T>(IEnumerable<T> values, bool strict) where T : struct, IComparable, IFormattable
	{
		foreach (T value in values)
		{
			Write(value, strict);
		}
	}

	public override void Write(short value)
	{
		if (NeedsReversion)
		{
			byte[] bytes = BitConverter.GetBytes(value);
			WriteReversed(bytes);
		}
		else
		{
			base.Write(value);
		}
	}

	public void Write(IEnumerable<short> values)
	{
		foreach (short value in values)
		{
			Write(value);
		}
	}

	public override void Write(int value)
	{
		if (NeedsReversion)
		{
			byte[] bytes = BitConverter.GetBytes(value);
			WriteReversed(bytes);
		}
		else
		{
			base.Write(value);
		}
	}

	public void Write(IEnumerable<int> values)
	{
		foreach (int value in values)
		{
			Write(value);
		}
	}

	public override void Write(long value)
	{
		if (NeedsReversion)
		{
			byte[] bytes = BitConverter.GetBytes(value);
			WriteReversed(bytes);
		}
		else
		{
			base.Write(value);
		}
	}

	public void Write(IEnumerable<long> values)
	{
		foreach (long value in values)
		{
			Write(value);
		}
	}

	public void WriteObject(object value)
	{
		if (value != null)
		{
			WriteObject(null, BinaryMemberAttribute.Default, value.GetType(), value);
		}
	}

	public override void Write(float value)
	{
		if (NeedsReversion)
		{
			byte[] bytes = BitConverter.GetBytes(value);
			WriteReversed(bytes);
		}
		else
		{
			base.Write(value);
		}
	}

	public void Write(IEnumerable<float> values)
	{
		foreach (float value in values)
		{
			Write(value);
		}
	}

	public void Write(string value, BinaryStringFormat format)
	{
		Write(value, format, Encoding);
	}

	public void Write(string value, BinaryStringFormat format, Encoding encoding)
	{
		switch (format)
		{
		case BinaryStringFormat.ByteLengthPrefix:
			WriteByteLengthPrefixString(value, encoding);
			break;
		case BinaryStringFormat.WordLengthPrefix:
			WriteWordLengthPrefixString(value, encoding);
			break;
		case BinaryStringFormat.DwordLengthPrefix:
			WriteDwordLengthPrefixString(value, encoding);
			break;
		case BinaryStringFormat.VariableLengthPrefix:
			WriteVariableLengthPrefixString(value, encoding);
			break;
		case BinaryStringFormat.ZeroTerminated:
			WriteZeroTerminatedString(value, encoding);
			break;
		case BinaryStringFormat.NoPrefixOrTermination:
			WriteNoPrefixOrTerminationString(value, encoding);
			break;
		default:
			throw new ArgumentOutOfRangeException("format", "The specified binary string format is invalid.");
		}
	}

	public void Write(IEnumerable<string> values)
	{
		foreach (string value in values)
		{
			Write(value);
		}
	}

	public void Write(IEnumerable<string> values, BinaryStringFormat format)
	{
		foreach (string value in values)
		{
			Write(value, format);
		}
	}

	public void Write(IEnumerable<string> values, BinaryStringFormat format, Encoding encoding)
	{
		foreach (string value in values)
		{
			Write(value, format, encoding);
		}
	}

	public override void Write(ushort value)
	{
		if (NeedsReversion)
		{
			byte[] bytes = BitConverter.GetBytes(value);
			WriteReversed(bytes);
		}
		else
		{
			base.Write(value);
		}
	}

	public void Write(IEnumerable<ushort> values)
	{
		foreach (ushort value in values)
		{
			Write(value);
		}
	}

	public override void Write(uint value)
	{
		if (NeedsReversion)
		{
			byte[] bytes = BitConverter.GetBytes(value);
			WriteReversed(bytes);
		}
		else
		{
			base.Write(value);
		}
	}

	public void Write(IEnumerable<uint> values)
	{
		foreach (uint value in values)
		{
			Write(value);
		}
	}

	public override void Write(ulong value)
	{
		if (NeedsReversion)
		{
			byte[] bytes = BitConverter.GetBytes(value);
			WriteReversed(bytes);
		}
		else
		{
			base.Write(value);
		}
	}

	public void Write(IEnumerable<ulong> values)
	{
		foreach (ulong value in values)
		{
			Write(value);
		}
	}

	private void WriteReversed(byte[] bytes)
	{
		Array.Reverse(bytes);
		base.Write(bytes);
	}

	private byte[] DecimalToBytes(decimal value)
	{
		byte[] array = new byte[16];
		Buffer.BlockCopy(decimal.GetBits(value), 0, array, 0, 16);
		return array;
	}

	private void WriteEnum(Type type, object value, bool strict)
	{
		if (strict && !EnumExtensions.IsValid(type, value))
		{
			throw new InvalidDataException($"Value {value} to write is not defined in the given enum type {type}.");
		}
		WriteObject(null, BinaryMemberAttribute.Default, Enum.GetUnderlyingType(type), value);
	}

	private void WriteObject(object instance, BinaryMemberAttribute attribute, Type type, object value)
	{
		if (attribute.Converter == null)
		{
			if (value == null)
			{
				return;
			}
			if (type == typeof(string))
			{
				Write((string)value, attribute.StringFormat);
				return;
			}
			if (type.TryGetEnumerableElementType(out var elementType))
			{
				foreach (object item in (IEnumerable)value)
				{
					WriteObject(null, BinaryMemberAttribute.Default, elementType, item);
				}
				return;
			}
			if (type == typeof(bool))
			{
				Write((bool)value, attribute.BooleanFormat);
			}
			else if (type == typeof(byte))
			{
				Write((byte)value);
			}
			else if (type == typeof(DateTime))
			{
				Write((DateTime)value, attribute.DateTimeFormat);
			}
			else if (type == typeof(decimal))
			{
				Write((decimal)value);
			}
			else if (type == typeof(double))
			{
				Write((double)value);
			}
			else if (type == typeof(short))
			{
				Write((short)value);
			}
			else if (type == typeof(int))
			{
				Write((int)value);
			}
			else if (type == typeof(long))
			{
				Write((long)value);
			}
			else if (type == typeof(sbyte))
			{
				Write((sbyte)value);
			}
			else if (type == typeof(float))
			{
				Write((float)value);
			}
			else if (type == typeof(ushort))
			{
				Write((ushort)value);
			}
			else if (type == typeof(uint))
			{
				Write((uint)value);
			}
			else if (type == typeof(ulong))
			{
				Write((uint)value);
			}
			else if (type.GetTypeInfo().IsEnum)
			{
				WriteEnum(type, value, attribute.Strict);
			}
			else
			{
				WriteCustomObject(type, value, Position);
			}
		}
		else
		{
			BinaryConverterCache.GetConverter(attribute.Converter).Write(this, instance, attribute, value);
		}
	}

	private void WriteCustomObject(Type type, object instance, long startOffset)
	{
		TypeData typeData = TypeData.GetTypeData(type);
		if (typeData.Attribute.Inherit && typeData.TypeInfo.BaseType != null)
		{
			WriteCustomObject(typeData.TypeInfo.BaseType, instance, startOffset);
		}
		foreach (MemberData member in typeData.Members)
		{
			if (member.Attribute.OffsetOrigin == OffsetOrigin.Begin)
			{
				Position = startOffset + member.Attribute.Offset;
			}
			else
			{
				Position += member.Attribute.Offset;
			}
			MemberInfo memberInfo = member.MemberInfo;
			if ((object)memberInfo != null)
			{
				object value;
				if (!(memberInfo is FieldInfo fieldInfo))
				{
					if (!(memberInfo is PropertyInfo propertyInfo))
					{
						goto IL_00cc;
					}
					value = propertyInfo.GetValue(instance);
				}
				else
				{
					value = fieldInfo.GetValue(instance);
				}
				if (member.Type.GetEnumerableElementType() == null)
				{
					WriteObject(instance, member.Attribute, member.Type, value);
					continue;
				}
				foreach (object item in (IEnumerable)value)
				{
					WriteObject(instance, member.Attribute, member.Type, item);
				}
				continue;
			}
			goto IL_00cc;
			IL_00cc:
			throw new InvalidOperationException($"Tried to write an invalid member {member.MemberInfo}.");
		}
	}

	private void WriteByteLengthPrefixString(string value, Encoding encoding)
	{
		Write((byte)value.Length);
		Write(encoding.GetBytes(value));
	}

	private void WriteWordLengthPrefixString(string value, Encoding encoding)
	{
		Write((short)value.Length);
		Write(encoding.GetBytes(value));
	}

	private void WriteDwordLengthPrefixString(string value, Encoding encoding)
	{
		Write(value.Length);
		Write(encoding.GetBytes(value));
	}

	private void WriteVariableLengthPrefixString(string value, Encoding encoding)
	{
		Write7BitEncodedInt(value.Length);
		Write(encoding.GetBytes(value));
	}

	private void WriteZeroTerminatedString(string value, Encoding encoding)
	{
		Write(encoding.GetBytes(value));
		Write((byte)0);
	}

	private void WriteNoPrefixOrTerminationString(string value, Encoding encoding)
	{
		Write(encoding.GetBytes(value));
	}
}

File: src/Starfield2026.MiniToolbox/src/MiniToolbox.Api/src/bntx/BinaryData/BinaryDataWriter.ts
import { ByteOrder } from './ByteOrder.js';
import { ByteOrderHelper } from './ByteOrderHelper.js';
import { BinaryBooleanFormat } from './BinaryBooleanFormat.js';
import { BinaryDateTimeFormat } from './BinaryDateTimeFormat.js';
import { BinaryStringFormat } from './BinaryStringFormat.js';
import { BinaryMemberAttribute } from './BinaryMemberAttribute.js';
import { BinaryConverterCache } from './BinaryConverterCache.js';
import { TypeData } from './TypeData.js';
import { OffsetOrigin } from './OffsetOrigin.js';
import { EnumExtensions } from './EnumExtensions.js';
import { Offset } from './Offset.js';
import { SeekTask } from './SeekTask.js';

/**
 * 1:1 port of BinaryDataWriter.cs
 * Full-featured binary writer with endianness support, object serialization,
 * and all the string/boolean/datetime format variants.
 */
export class BinaryDataWriter {
    private _buffer: Buffer;
    private _position: number = 0;
    private _byteOrder: ByteOrder;
    private _needsReversion: boolean = false;
    private _capacity: number;

    get byteOrder(): ByteOrder { return this._byteOrder; }
    set byteOrder(value: ByteOrder) {
        this._byteOrder = value;
        this._needsReversion = this._byteOrder !== ByteOrderHelper.systemByteOrder;
    }

    get needsReversion(): boolean { return this._needsReversion; }

    get position(): number { return this._position; }
    set position(value: number) { this._position = value; }

    get length(): number { return this._position; }

    constructor(initialCapacity: number = 4096) {
        this._capacity = initialCapacity;
        this._buffer = Buffer.alloc(this._capacity);
        this._byteOrder = ByteOrderHelper.systemByteOrder;
    }

    /** Get the written data as a Buffer */
    toBuffer(): Buffer {
        return this._buffer.subarray(0, this._position);
    }

    private ensureCapacity(additional: number): void {
        while (this._position + additional > this._capacity) {
            this._capacity *= 2;
            const newBuf = Buffer.alloc(this._capacity);
            this._buffer.copy(newBuf);
            this._buffer = newBuf;
        }
    }

    // ── Alignment ──

    align(alignment: number): void {
        this.seek((-this._position % alignment + alignment) % alignment);
    }

    reserveOffset(): Offset {
        return new Offset(this);
    }

    reserveOffsets(count: number): Offset[] {
        const offsets: Offset[] = [];
        for (let i = 0; i < count; i++) {
            offsets.push(this.reserveOffset());
        }
        return offsets;
    }

    // ── Seek ──

    seek(offset: number, origin?: 'begin' | 'current' | 'end'): number {
        const o = origin ?? 'current';
        switch (o) {
            case 'begin': this._position = offset; break;
            case 'current': this._position += offset; break;
            case 'end': this._position = this._buffer.length + offset; break;
        }
        return this._position;
    }

    temporarySeek(offset?: number, origin?: 'begin' | 'current' | 'end'): SeekTask {
        return new SeekTask(this, offset ?? 0, origin ?? 'current');
    }

    // ── Boolean ──

    writeBoolean(value: boolean, format?: BinaryBooleanFormat): void {
        const fmt = format ?? BinaryBooleanFormat.NonZeroByte;
        switch (fmt) {
            case BinaryBooleanFormat.NonZeroByte:
                this.writeByte(value ? 1 : 0);
                break;
            case BinaryBooleanFormat.NonZeroWord:
                this.writeInt16(value ? 1 : 0);
                break;
            case BinaryBooleanFormat.NonZeroDword:
                this.writeInt32(value ? 1 : 0);
                break;
            default:
                throw new Error('The specified binary boolean format is invalid.');
        }
    }

    writeBooleans(values: boolean[], format?: BinaryBooleanFormat): void {
        for (const v of values) this.writeBoolean(v, format);
    }

    // ── DateTime ──

    writeDateTime(value: Date, format?: BinaryDateTimeFormat): void {
        const fmt = format ?? BinaryDateTimeFormat.NetTicks;
        switch (fmt) {
            case BinaryDateTimeFormat.CTime:
                this.writeUInt32(Math.floor(value.getTime() / 1000));
                break;
            case BinaryDateTimeFormat.NetTicks: {
                const epochTicks = BigInt('621355968000000000');
                const ticks = epochTicks + BigInt(value.getTime()) * BigInt(10000);
                this.writeInt64(Number(ticks));
                break;
            }
            default:
                throw new Error('The specified binary date time format is invalid.');
        }
    }

    writeDateTimes(values: Date[], format?: BinaryDateTimeFormat): void {
        for (const v of values) this.writeDateTime(v, format);
    }

    // ── Primitives ──

    writeByte(value: number): void {
        this.ensureCapacity(1);
        this._buffer.writeUInt8(value, this._position);
        this._position += 1;
    }

    writeBytes(values: Buffer | number[]): void {
        const buf = Buffer.isBuffer(values) ? values : Buffer.from(values);
        this.ensureCapacity(buf.length);
        buf.copy(this._buffer, this._position);
        this._position += buf.length;
    }

    writeSByte(value: number): void {
        this.ensureCapacity(1);
        this._buffer.writeInt8(value, this._position);
        this._position += 1;
    }

    writeInt16(value: number): void {
        this.ensureCapacity(2);
        if (this._needsReversion) {
            this._buffer.writeInt16BE(value, this._position);
        } else {
            this._buffer.writeInt16LE(value, this._position);
        }
        this._position += 2;
    }

    writeInt16s(values: number[]): void {
        for (const v of values) this.writeInt16(v);
    }

    writeUInt16(value: number): void {
        this.ensureCapacity(2);
        if (this._needsReversion) {
            this._buffer.writeUInt16BE(value, this._position);
        } else {
            this._buffer.writeUInt16LE(value, this._position);
        }
        this._position += 2;
    }

    writeUInt16s(values: number[]): void {
        for (const v of values) this.writeUInt16(v);
    }

    writeInt32(value: number): void {
        this.ensureCapacity(4);
        if (this._needsReversion) {
            this._buffer.writeInt32BE(value, this._position);
        } else {
            this._buffer.writeInt32LE(value, this._position);
        }
        this._position += 4;
    }

    writeInt32s(values: number[]): void {
        for (const v of values) this.writeInt32(v);
    }

    writeUInt32(value: number): void {
        this.ensureCapacity(4);
        if (this._needsReversion) {
            this._buffer.writeUInt32BE(value, this._position);
        } else {
            this._buffer.writeUInt32LE(value, this._position);
        }
        this._position += 4;
    }

    writeUInt32s(values: number[]): void {
        for (const v of values) this.writeUInt32(v);
    }

    writeInt64(value: number): void {
        this.ensureCapacity(8);
        if (this._needsReversion) {
            this._buffer.writeBigInt64BE(BigInt(value), this._position);
        } else {
            this._buffer.writeBigInt64LE(BigInt(value), this._position);
        }
        this._position += 8;
    }

    writeInt64s(values: number[]): void {
        for (const v of values) this.writeInt64(v);
    }

    writeUInt64(value: number): void {
        this.ensureCapacity(8);
        if (this._needsReversion) {
            this._buffer.writeBigUInt64BE(BigInt(value), this._position);
        } else {
            this._buffer.writeBigUInt64LE(BigInt(value), this._position);
        }
        this._position += 8;
    }

    writeUInt64s(values: number[]): void {
        for (const v of values) this.writeUInt64(v);
    }

    writeSingle(value: number): void {
        this.ensureCapacity(4);
        if (this._needsReversion) {
            this._buffer.writeFloatBE(value, this._position);
        } else {
            this._buffer.writeFloatLE(value, this._position);
        }
        this._position += 4;
    }

    writeSingles(values: number[]): void {
        for (const v of values) this.writeSingle(v);
    }

    writeDouble(value: number): void {
        this.ensureCapacity(8);
        if (this._needsReversion) {
            this._buffer.writeDoubleBE(value, this._position);
        } else {
            this._buffer.writeDoubleLE(value, this._position);
        }
        this._position += 8;
    }

    writeDoubles(values: number[]): void {
        for (const v of values) this.writeDouble(v);
    }

    // ── String ──

    writeString(value: string, format?: BinaryStringFormat): void {
        const fmt = format ?? BinaryStringFormat.VariableLengthPrefix;
        const encoded = Buffer.from(value, 'utf8');
        switch (fmt) {
            case BinaryStringFormat.ByteLengthPrefix:
                this.writeByte(value.length);
                this.writeBytes(encoded);
                break;
            case BinaryStringFormat.WordLengthPrefix:
                this.writeInt16(value.length);
                this.writeBytes(encoded);
                break;
            case BinaryStringFormat.DwordLengthPrefix:
                this.writeInt32(value.length);
                this.writeBytes(encoded);
                break;
            case BinaryStringFormat.VariableLengthPrefix:
                this.write7BitEncodedInt(value.length);
                this.writeBytes(encoded);
                break;
            case BinaryStringFormat.ZeroTerminated:
                this.writeBytes(encoded);
                this.writeByte(0);
                break;
            case BinaryStringFormat.NoPrefixOrTermination:
                this.writeBytes(encoded);
                break;
            default:
                throw new Error('The specified binary string format is invalid.');
        }
    }

    writeStrings(values: string[], format?: BinaryStringFormat): void {
        for (const v of values) this.writeString(v, format);
    }

    // ── Enum ──

    writeEnum(enumObj: Record<string, number | string>, value: number, strict: boolean = false): void {
        if (strict && !EnumExtensions.isValid(enumObj, value)) {
            throw new Error(`Value ${value} to write is not defined in the given enum type.`);
        }
        this.writeInt32(value);
    }

    // ── Object (reflection-based serialization) ──

    writeObject(value: any, typeName?: string): void {
        if (value === null || value === undefined) return;
        const name = typeName ?? 'object';
        this.writeObjectInternal(null, BinaryMemberAttribute.Default, name, value);
    }

    private writeObjectInternal(instance: any, attribute: BinaryMemberAttribute, typeName: string, value: any): void {
        if (attribute.converter !== null) {
            BinaryConverterCache.getConverter(attribute.converter).write(this, instance, attribute, value);
            return;
        }

        if (value === null || value === undefined) return;

        switch (typeName) {
            case 'string': this.writeString(value, attribute.stringFormat); return;
            case 'boolean': this.writeBoolean(value, attribute.booleanFormat); return;
            case 'byte': this.writeByte(value); return;
            case 'sbyte': this.writeSByte(value); return;
            case 'int16': this.writeInt16(value); return;
            case 'uint16': this.writeUInt16(value); return;
            case 'int32': this.writeInt32(value); return;
            case 'uint32': this.writeUInt32(value); return;
            case 'int64': this.writeInt64(value); return;
            case 'uint64': this.writeUInt64(value); return;
            case 'float': this.writeSingle(value); return;
            case 'double': this.writeDouble(value); return;
            case 'datetime': this.writeDateTime(value, attribute.dateTimeFormat); return;
            default:
                if (Array.isArray(value)) {
                    const elementType = typeName.replace('[]', '');
                    for (const item of value) {
                        this.writeObjectInternal(null, BinaryMemberAttribute.Default, elementType, item);
                    }
                    return;
                }
                this.writeCustomObject(typeName, value, this._position);
        }
    }

    private writeCustomObject(typeName: string, instance: any, startOffset: number): void {
        const typeData = TypeData.getTypeData(typeName);

        for (const member of typeData.members) {
            if (member.attribute.offsetOrigin === OffsetOrigin.Begin) {
                this._position = startOffset + member.attribute.offset;
            } else if (member.attribute.offset !== 0) {
                this._position += member.attribute.offset;
            }

            const value = member.getValue(instance);

            if (member.type.endsWith('[]') && Array.isArray(value)) {
                const elementType = member.type.replace('[]', '');
                for (const item of value) {
                    this.writeObjectInternal(instance, member.attribute, elementType, item);
                }
            } else {
                this.writeObjectInternal(instance, member.attribute, member.type, value);
            }
        }
    }

    // ── Private helpers ──

    private write7BitEncodedInt(value: number): void {
        let v = value;
        while (v >= 0x80) {
            this.writeByte((v & 0x7F) | 0x80);
            v >>= 7;
        }
        this.writeByte(v);
    }
}
